libcodechecker/analyze/store_handler.py

```diff
@@ -228,25 +234,45 @@ def addCheckerRun(session, storage_session, command, name, tag, username,
 
             LOG.info('Removing previous analysis results ...')
             session.delete(run)
+            session.flush()
 
+            # The adding of the new run must happen on a separate session,
+            # so that the lock is committed to the database.
             checker_run = Run(name, version, command)
-            session.add(checker_run)
-            session.flush()
-            run_id = checker_run.id
+            checker_run.lock_timestamp = datetime.now()
+            session2.add(checker_run)
+            session2.commit()
+            session2.close()
 
+            run_id = checker_run.id
         elif run:
             # There is already a run, update the results.
+
+            # Mark the run locked and save this information into the database.
+            run2.lock_timestamp = datetime.now()
+            session2.commit()
+            session2.close()
+
+            # Update the run's other metadata, but these only save into the
+            # transaction's buffer, so if the store fails, the previous
+            # timestamp and duration is recovered.
             run.date = datetime.now()
             run.command = command
             run.duration = -1
             session.flush()
             run_id = run.id
         else:
-            # There is no run create new.
+            # There is no run, create new and save into the database.
+            # This is done on the second transaction, so the first transaction
+            # keeps the run's related (but not yet saved) data only in the
+            # transaction buffer.
             checker_run = Run(name, version, command)
-            session.add(checker_run)
-            session.flush()
+            checker_run.lock_timestamp = datetime.now()
+            session2.add(checker_run)
+            session2.commit()
+
             run_id = checker_run.id
+            session2.close()
```

Comment:
There is some code duplication. Maybe the could could be rewritten to delete the run if force was provided and then fall back to this codepath?